socks5/socks5.go

```diff
@@ -0,0 +1,273 @@
+			2, // number of authentication methods
+			authNone,
+			authUsernamePassword,
+		}
+		_, err = conn.Write(buf)
+	} else {
+		buf := []byte{
+			protocolVersion,
+			1, // number of authencation methods
+			authNone,
+		}
+		_, err = conn.Write(buf)
+	}
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	var res [2]byte
+	if _, err := io.ReadFull(conn, res[:]); err != nil {
+		conn.Close()
+		return nil, err
+	}
+	if res[0] != protocolVersion {
+		conn.Close()
+		return nil, ErrInvalidProxyResponse
+	}
+
+	switch res[1] {
+	case authGssApi:
+		err = ErrNoAcceptableAuthMethod
+	case authNone:
+		// Do nothing
+	case authUnavailable:
+		err = ErrNoAcceptableAuthMethod
+	case authUsernamePassword:
+		buf := make([]byte, 3+len(user)+len(pass))
+		buf[0] = 1 // version
+		buf[1] = byte(len(user))
+		copy(buf[2:], user)
+		buf[2+len(user)] = byte(len(pass))
+		copy(buf[2+len(user)+1:], pass)
+		if _, err = conn.Write(buf); err != nil {
+			conn.Close()
+			return nil, err
+		}
+		if _, err = io.ReadFull(conn, buf[:2]); err != nil {
+			conn.Close()
+			return nil, err
+		}
+		if buf[0] != 1 { // version
+			err = ErrInvalidProxyResponse
+		} else if buf[1] != 0 { // 0 = success
+			err = ErrAuthFailed
+		}
+	default:
+		err = ErrInvalidProxyResponse
+	}
+
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	// Command and connection request.
+	buf := make([]byte, 5+len(remoteHost)+2)
+	buf[0] = protocolVersion
+	buf[1] = commandTcpConnect
+	buf[2] = 0 // reserved
+	buf[3] = addressTypeDomain
+	buf[4] = byte(len(remoteHost))
+	copy(buf[5:], remoteHost)
+	buf[5+len(remoteHost)] = byte(remotePort >> 8)
+	buf[6+len(remoteHost)] = byte(remotePort & 0xff)
+	if _, err := conn.Write(buf); err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	// Server response
+	var res2 [4]byte
+	if _, err := io.ReadFull(conn, res2[:]); err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	if res2[0] != protocolVersion {
+		conn.Close()
+		return nil, ErrInvalidProxyResponse
+	}
+	if res2[1] != statusRequestGranted {
+		conn.Close()
+		err := statusErrors[res2[1]]
+		if err == nil {
+			err = ErrInvalidProxyResponse
+		}
+		return nil, err
+	}
+
+	switch res2[3] {
+	case addressTypeDomain:
+		var domainLen [1]byte
+		if _, err := io.ReadFull(conn, domainLen[:]); err != nil {
+			conn.Close()
+			return nil, err
+		}
+		b := make([]byte, int(domainLen[0]))
+		if _, err := io.ReadFull(conn, b); err != nil {
+			conn.Close()
+			return nil, err
+		}
+	case addressTypeIPv4:
+		var b [4]byte
+		if _, err := io.ReadFull(conn, b[:]); err != nil {
+			conn.Close()
+			return nil, err
+		}
+	case addressTypeIPv6:
+		var b [16]byte
+		if _, err := io.ReadFull(conn, b[:]); err != nil {
+			conn.Close()
+			return nil, err
+		}
+	default:
+		conn.Close()
+		return nil, ErrInvalidProxyResponse
+	}
+
+	if _, err := io.ReadFull(conn, res2[:2]); err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+func (s *Socks5) Randomize() bool {
```

Comment:
Should probably used `Randomized`.  This reads like a verb as if it's going to do something rather than returning the state of randomization.